File: app/src/main/java/io/temco/guhada/data/viewmodel/account/JoinViewModel.java
package io.temco.guhada.data.viewmodel.account;

import androidx.databinding.Bindable;
import androidx.databinding.ObservableBoolean;

import java.util.Observable;
import java.util.Observer;

import io.temco.guhada.BR;
import io.temco.guhada.R;
import io.temco.guhada.common.BaseApplication;
import io.temco.guhada.common.listener.OnJoinListener;
import io.temco.guhada.common.util.CommonUtil;
import io.temco.guhada.common.util.CustomLog;
import io.temco.guhada.data.model.base.BaseModel;
import io.temco.guhada.data.model.user.User;
import io.temco.guhada.data.server.UserServer;
import io.temco.guhada.data.viewmodel.base.BaseObservableViewModel;

import static android.app.Activity.RESULT_CANCELED;
import static android.app.Activity.RESULT_OK;

public class JoinViewModel extends BaseObservableViewModel implements Observer {
    private String toolBarTitle = "";
    private User user = new User();
    private OnJoinListener listener;

    // TERMS CHECKED
    private ObservableBoolean allChecked = new ObservableBoolean(false);
    private ObservableBoolean essentialChecked = new ObservableBoolean(false);
    private ObservableBoolean optionalChecked = new ObservableBoolean(false);

    public JoinViewModel(OnJoinListener listener) {
        this.listener = listener;
        user.addObserver(this);
    }

    // GETTER & SETTER
    @Bindable
    public String getToolBarTitle() {
        return toolBarTitle;
    }

    public void setToolBarTitle(String toolBarTitle) {
        this.toolBarTitle = toolBarTitle;
    }

    @Bindable
    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }

    @Bindable
    public ObservableBoolean getAllChecked() {
        return allChecked;
    }

    public void setAllChecked(ObservableBoolean allChecked) {
        this.allChecked = allChecked;
    }

    @Bindable
    public ObservableBoolean getEssentialChecked() {
        return essentialChecked;
    }

    public void setEssentialChecked(boolean flag) {
        this.essentialChecked.set(flag);
        notifyPropertyChanged(BR.essentialChecked);
    }

    @Bindable
    public ObservableBoolean getOptionalChecked() {
        return optionalChecked;
    }


    // CLICK LISTENER
    public void onClickBack() {
        listener.closeActivity(RESULT_CANCELED);
    }

    public void onClickSignUp() {
        if (!CommonUtil.validateEmail(user.getEmail())) {
            listener.showSnackBar(BaseApplication.getInstance().getResources().getString(R.string.join_message_wrongidformat));
        } else if (user.getPassword().length() < 8 || user.getPassword().length() > 15) {
            listener.showSnackBar(BaseApplication.getInstance().getResources().getString(R.string.join_message_wrondpwdlength));
        } else if (!CommonUtil.validatePassword(user.getPassword())) {
            listener.showSnackBar(BaseApplication.getInstance().getResources().getString(R.string.join_message_wrongpwdformat));
        } else if (!user.getConfirmPassword().equals(user.getPassword())) {
            listener.showSnackBar(BaseApplication.getInstance().getResources().getString(R.string.join_message_notequalpwd));
        } else {
            user.deleteObserver(this);
            UserServer.signUp((success, o) -> {
                if (success) {
                    BaseModel model = ((BaseModel) o);
                    switch (model.resultCode) {
                        case 200:
                            String email = model.data instanceof String ? (String) model.data : user.getEmail();
                            user.setEmail(email);
                            // listener.showMessage(email + " 가입 완료");
                            listener.closeActivity(RESULT_OK);
                            return;
                        case 6001: // ALREADY EXIST EMAIL
                            listener.showSnackBar(BaseApplication.getInstance().getResources().getString(R.string.join_message_existemail));
                            break;
                        case 6002: // INVALID PASSWORD
                            listener.showSnackBar(BaseApplication.getInstance().getResources().getString(R.string.join_message_wrongpwdformat));
                            break;
                    }
                } else {
                    listener.showSnackBar((String) o);
                }
            }, user);
        }
    }

    // TERMS CHECK LISTENER
    public void onCheckAll(boolean checked) {
        if (checked) {
            user.setAgreePurchaseTos(true);
            user.setAgreeCollectPersonalInfoTos(true);
            user.setAgreeSaleTos(true);
            user.setAgreeEmailReception(true);
            user.setAgreeSmsReception(true);
        } else {
            if (essentialChecked.get() && optionalChecked.get()) {
                user.setAgreePurchaseTos(false);
                user.setAgreeCollectPersonalInfoTos(false);
                user.setAgreeSaleTos(false);
                user.setAgreeEmailReception(false);
                user.setAgreeSmsReception(false);
            }
        }
    }

    public void onCheckEssential(boolean checked) {
        if (checked) {
            essentialChecked = new ObservableBoolean(true);
            user.setAgreePurchaseTos(true);
            user.setAgreeCollectPersonalInfoTos(true);
            notifyPropertyChanged(BR.essentialChecked);
        } else {
            if (user.getAgreePurchaseTos() && user.getAgreeCollectPersonalInfoTos()) {
                essentialChecked = new ObservableBoolean(true);
                user.setAgreePurchaseTos(false);
                user.setAgreeCollectPersonalInfoTos(false);
                notifyPropertyChanged(BR.essentialChecked);
            } else {
                essentialChecked = new ObservableBoolean(false);
            }
        }
    }

    public void onCheckOption(boolean checked) {
        if (checked) {
            optionalChecked = new ObservableBoolean(true);
            user.setAgreeSaleTos(true);
            user.setAgreeEmailReception(true);
            user.setAgreeSmsReception(true);
            notifyPropertyChanged(BR.optionalChecked);
        } else {
            if (user.getAgreeSaleTos() && user.getAgreeEmailReception() && user.getAgreeSmsReception()) {
                optionalChecked = new ObservableBoolean(false);
                user.setAgreeSaleTos(false);
                user.setAgreeEmailReception(false);
                user.setAgreeSmsReception(false);
                notifyPropertyChanged(BR.optionalChecked);
            } else {
                optionalChecked = new ObservableBoolean(false);
            }

        }
    }

    public void onCheckPurchaseTos(boolean checked) {
        user.setAgreePurchaseTos(checked);
    }

    public void onCheckPrivacyTos(boolean checked) {
        user.setAgreeCollectPersonalInfoTos(checked);
    }

    public void onCheckSaleTos(boolean checked) {
        user.setAgreeSaleTos(checked);
    }

    public void onCheckEmailReception(boolean checked) {
        user.setAgreeEmailReception(checked);
    }

    public void onCheckSmsReception(boolean checked) {
        user.setAgreeSmsReception(checked);
    }

    @Override
    public void update(Observable o, Object arg) {
        if (arg instanceof String) {
            switch ((String) arg) {
                // ESSENTIAL TERMS
                case "agreeCollectPersonalInfoTos":
                case "agreePurchaseTos":
                    boolean personalInfoTosChecked = user.getAgreeCollectPersonalInfoTos();
                    boolean purchaseTosChecked = user.getAgreePurchaseTos();
                    boolean isEssentialAllChecked = personalInfoTosChecked && purchaseTosChecked;
                    if (CustomLog.INSTANCE.getFlag())
                        CustomLog.INSTANCE.L("update", "isEssentialAllChecked", isEssentialAllChecked);
                    if (CustomLog.INSTANCE.getFlag())
                        CustomLog.INSTANCE.L("update", "essentialChecked.get() ", essentialChecked.get());
                    if (essentialChecked.get() != isEssentialAllChecked) {
                        essentialChecked = new ObservableBoolean(isEssentialAllChecked);
                        notifyPropertyChanged(BR.essentialChecked);
                    }
                    break;
                case "agreeSaleTos":
                case "agreeEmailReception":
                case "agreeSmsReception":
                    boolean saleTosChecked = user.getAgreeSaleTos();
                    boolean emailReceptionChecked = user.getAgreeEmailReception();
                    boolean smsReceptionChecked = user.getAgreeSmsReception();
                    boolean isOptionAllChecked = saleTosChecked && emailReceptionChecked && smsReceptionChecked;

                    if (optionalChecked.get() != isOptionAllChecked) {
                        optionalChecked = new ObservableBoolean(isOptionAllChecked);
                        notifyPropertyChanged(BR.optionalChecked);
                    }
                    break;
            }
        }
    }
}

File: app/src/main/java/io/temco/guhada/data/viewmodel/account/JoinViewModel.kt
package io.temco.guhada.data.viewmodel.account

import android.app.Activity.RESULT_CANCELED
import android.app.Activity.RESULT_OK
import androidx.databinding.Bindable
import androidx.databinding.ObservableBoolean
import io.temco.guhada.BR
import io.temco.guhada.R
import io.temco.guhada.common.BaseApplication
import io.temco.guhada.common.enum.ResultCode
import io.temco.guhada.common.listener.OnJoinListener
import io.temco.guhada.common.listener.OnServerListener
import io.temco.guhada.common.util.CommonUtil
import io.temco.guhada.common.util.CustomLog
import io.temco.guhada.data.model.base.BaseModel
import io.temco.guhada.data.model.user.User
import io.temco.guhada.data.server.UserServer
import io.temco.guhada.data.viewmodel.base.BaseObservableViewModel
import java.util.*

class JoinViewModel : BaseObservableViewModel(), Observer {
    lateinit var listener: OnJoinListener
    var toolBarTitle = ""
    var user = User()

    // TERMS CHECKED
    var allChecked = ObservableBoolean(false)
        @Bindable
        get() = field
    var essentialChecked = ObservableBoolean(false)
        @Bindable
        get() = field
    var optionalChecked = ObservableBoolean(false)
        @Bindable
        get() = field
    val selectTermPlusView = ObservableBoolean(false)
        @Bindable
        get() = field

    init {
        this.user.addObserver(this)
    }

    // CLICK LISTENER
    fun onClickBack() = listener.closeActivity(RESULT_CANCELED)

    fun onClickSignUp() {
        if (!CommonUtil.validateEmail(user.email))
            listener.showSnackBar(BaseApplication.getInstance().resources.getString(R.string.join_message_wrongidformat))
        else if (user.password.length < 8 || user.password.length > 15)
            listener.showSnackBar(BaseApplication.getInstance().resources.getString(R.string.join_message_wrondpwdlength))
        else if (!CommonUtil.validatePassword(user.password))
            listener.showSnackBar(BaseApplication.getInstance().resources.getString(R.string.join_message_wrongpwdformat))
        else if (user.confirmPassword != user.password)
            listener.showSnackBar(BaseApplication.getInstance().resources.getString(R.string.join_message_notequalpwd))
        else signUp()
    }

    private fun signUp() {
        UserServer.signUp(OnServerListener { success, o ->
            if (success) {
                val model = o as BaseModel<*>
                when (model.resultCode) {
                    ResultCode.SUCCESS.flag -> {
                        user.deleteObserver(this)
                        user.email = if (model.data is String) model.data as String else user.email
                        listener.closeActivity(RESULT_OK)
                    }
                    ResultCode.ALREADY_EXIST_EMAIL.flag -> listener.showSnackBar(BaseApplication.getInstance().resources.getString(R.string.join_message_existemail))
                    ResultCode.SIGNUP_INVALID_PASSWORD.flag -> listener.showSnackBar(BaseApplication.getInstance().resources.getString(R.string.join_message_wrongpwdformat))
                }
            } else {
                listener.showSnackBar(o as String)
            }
        }, user)
    }

    // TERMS CHECK LISTENER
    fun onCheckAll(checked: Boolean) {
        if (checked) {
            setEssentialTerms(true)
            setOptionalTerms(true)
        } else if (essentialChecked.get() && optionalChecked.get()) {
            setEssentialTerms(false)
            setOptionalTerms(false)
        }
    }

    fun onCheckEssential(checked: Boolean) {
        essentialChecked = ObservableBoolean(checked)
        notifyPropertyChanged(BR.essentialChecked)

        if (checked)
            setEssentialTerms(true)
        else if (user.agreePurchaseTos!! && user.agreeCollectPersonalInfoTos!!)
            setEssentialTerms(false)
    }

    fun onCheckOption(checked: Boolean) {
        optionalChecked = ObservableBoolean(checked)
        notifyPropertyChanged(BR.optionalChecked)

        if (checked)
            setOptionalTerms(true)
        else if (user.agreeSaleTos!! && user.agreeEmailReception!! && user.agreeSmsReception!!)
            setOptionalTerms(false)
    }

    private fun setEssentialTerms(checked: Boolean) {
        user.agreePurchaseTos = checked
        user.agreeCollectPersonalInfoTos = checked
    }

    private fun setOptionalTerms(checked: Boolean) {
        user.agreeSaleTos = checked
        user.agreeEmailReception = checked
        user.agreeSmsReception = checked
    }

    fun onClickSelectTermPlusView() {
        this.selectTermPlusView.set(true)
        notifyPropertyChanged(BR.selectTermPlusView)
    }

    fun onCheckPurchaseTos(checked: Boolean) {
        user.agreePurchaseTos = checked
    }

    fun onCheckPrivacyTos(checked: Boolean) {
        user.agreeCollectPersonalInfoTos = checked
    }

    fun onCheckSaleTos(checked: Boolean) {
        user.agreeSaleTos = checked
    }

    fun onCheckEmailReception(checked: Boolean) {
        user.agreeEmailReception = checked
    }

    fun onCheckSmsReception(checked: Boolean) {
        user.agreeSmsReception = checked
    }

    override fun update(o: Observable?, arg: Any?) {
        if (arg is String) {
            when (arg) {
                // ESSENTIAL TERMS
                "agreeCollectPersonalInfoTos", "agreePurchaseTos" -> {
                    val personalInfoTosChecked = user.agreeCollectPersonalInfoTos!!
                    val purchaseTosChecked = user.agreePurchaseTos!!
                    val isEssentialAllChecked = personalInfoTosChecked && purchaseTosChecked

                    if (essentialChecked.get() != isEssentialAllChecked) {
                        essentialChecked = ObservableBoolean(isEssentialAllChecked)
                        notifyPropertyChanged(BR.essentialChecked)
                    }

                    if (CustomLog.flag) CustomLog.L("update", "isEssentialAllChecked", isEssentialAllChecked)
                    if (CustomLog.flag) CustomLog.L("update", "essentialChecked.get() ", essentialChecked.get())
                }

                "agreeSaleTos", "agreeEmailReception", "agreeSmsReception" -> {
                    val saleTosChecked = user.agreeSaleTos!!
                    val emailReceptionChecked = user.agreeEmailReception!!
                    val smsReceptionChecked = user.agreeSmsReception!!
                    val isOptionAllChecked = saleTosChecked && emailReceptionChecked && smsReceptionChecked

                    if (optionalChecked.get() != isOptionAllChecked) {
                        optionalChecked = ObservableBoolean(isOptionAllChecked)
                        notifyPropertyChanged(BR.optionalChecked)
                    }
                }
            }
        }
    }
}